homeassistant/components/smarttub/controller.py

```diff
@@ -0,0 +1,110 @@
+"""Interface to the SmartTub API."""
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from aiohttp import client_exceptions
+import async_timeout
+from smarttub import APIError, LoginFailed, SmartTub
+from smarttub.api import Account
+
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, POLLING_TIMEOUT, SCAN_INTERVAL
+from .helpers import get_spa_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmartTubController:
+    """Interface between Home Assistant and the SmartTub API."""
+
+    def __init__(self, hass):
+        """Initialize an interface to SmartTub."""
+        self._hass = hass
+        self._account = None
+        self.spas = set()
+        self._spa_devices = {}
+
+        self.coordinator = None
+
+    async def async_setup_entry(self, entry):
+        """Perform initial setup.
+
+        Authenticate, query static state, set up polling, and otherwise make
+        ready for normal operations .
+        """
+
+        try:
+            self._account = await self.login(
+                entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD]
+            )
+        except LoginFailed:
+            # credentials were changed or invalidated, we need new ones
+
+            return False
+        except (
+            asyncio.TimeoutError,
+            client_exceptions.ClientOSError,
+            client_exceptions.ServerDisconnectedError,
+            client_exceptions.ContentTypeError,
+        ) as err:
+            raise ConfigEntryNotReady from err
+
+        self.spas = await self._account.get_spas()
+
+        self.coordinator = DataUpdateCoordinator(
+            self._hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update_data,
+            update_interval=timedelta(seconds=SCAN_INTERVAL),
+        )
+
+        await self.coordinator.async_refresh()
+
+        await self.async_register_devices(entry)
+
+        return True
+
+    async def async_update_data(self):
+        """Query the API and return the new state."""
+
+        data = {}
+        try:
+            async with async_timeout.timeout(POLLING_TIMEOUT):
+                for spa in self.spas:
+                    data[spa.id] = {"status": await spa.get_status()}
```

Comment:
Can we get the status concurrently with `asyncio.gather` or do we need to do it sequentially?

Reply:
We could, though it's very unusual for an account to have more than one spa, so it didn't seem worth the complexity. Do you want me to do it anyway?

Reply:
And if there were a bunch of spas on one account, we wouldn't want to hammer the API. It has fairly restrictive rate limits.